modyn/supervisor/internal/pipeline_executor/evaluation_executor.py

```diff
@@ -292,40 +295,31 @@ def _single_evaluation(self, log: StageLog, eval_status_queue: Queue, eval_req:
                     self.grpc.init_evaluator()
                     raise e
 
+        def get_failure_reason(eval_aborted_reason: EvaluationAbortedReason) -> str:
+            return EvaluationAbortedReason.DESCRIPTOR.values_by_number[eval_aborted_reason].name
+
         if not response.evaluation_started:
-            log.info.failure_reason = EvaluationAbortedReason.DESCRIPTOR.values_by_number[
-                response.eval_aborted_reason
-            ].name
+            failure_reason = get_failure_reason(response.interval_responses[0].eval_aborted_reason)
             logger.error(
-                f"Evaluation for model {eval_req.id_model} on split {eval_req.interval_start} to "
-                f"{eval_req.interval_end} not started with reason: {log.info.failure_reason}."
+                f"Evaluation for model {model_id_to_eval} on split {interval_start} to "
+                f"{interval_end} not started with reason: {failure_reason}."
             )
-            return
+            return failure_reason, {}
 
-        logger.info(
-            f"Evaluation started for model {eval_req.id_model} on split {eval_req.interval_start} "
-            f"to {eval_req.interval_end}."
-        )
-        reporter = EvaluationStatusReporter(
-            eval_status_queue,
-            response.evaluation_id,
-            eval_req.dataset_id,
-            response.dataset_size,
-        )
-        evaluation = {response.evaluation_id: reporter}
-        reporter.create_tracker()
-        self.grpc.wait_for_evaluation_completion(eval_req.training_id, evaluation)
-
-        eval_result_writer = JsonResultWriter(self.pipeline_id, eval_req.trigger_id, self.pipeline_logdir)
-        self.grpc.store_evaluation_results([eval_result_writer], evaluation)
-        self.grpc.cleanup_evaluations([int(i) for i in evaluation])
-        assert isinstance(eval_result_writer, JsonResultWriter)
-
-        log.info.results = (
-            eval_result_writer.results["datasets"][0][dataset_config.dataset_id]
-            if eval_result_writer.results["datasets"]
-            else {}
-        )
+        logger.info(f"Evaluation started for model {model_id_to_eval} on split {interval_start} " f"to {interval_end}.")
+        self.grpc.wait_for_evaluation_completion(response.evaluation_id)
+        eval_data = self.grpc.get_evaluation_results(response.evaluation_id)
+        self.grpc.cleanup_evaluations([response.evaluation_id])
+        # here we assume only one interval is evaluated
+        eval_results: dict = {"dataset_size": response.interval_responses[0].dataset_size, "metrics": []}
+
+        for single_eval_data in eval_data:
```

Comment:
I think related to my other question on the interval_index I have a bigger question: It does not appear that we have easy access to the interval index here. Basically there is a 1:1 mapping between interval index and the interval on which we evaluate. Instead of somehow re-creating the mapping at the supervisor, it might be a lot easier to use a tuple (eval_interval_start, eval_interval_end) at the evaluator as the identifier of an interval, no? Because then we can just return that in the results instead of this index which we don't easily have access to. WDYT? If it's unclear, I will be at the office in ~1 hour so we can chat about it

Reply:
I see your point. My thought is that, later, this function will be rewritten so that it takes a list of intervals, packs them inside the `EvaluateModelRequest`, and then contact the evaluator to launch evaluation. So this function always has the full list of intervals when reading the response from the evaluator.

That's why I feel it is more convenient to give an index so here this `single_batched_evaluation` function can directly index to the original interval list. @robinholzi what's your thought here?

Reply:
Also, it is possible that (for whatever reason) the client sends a request with duplicate intervals inside, e.g. two [100, 200], in this case we cannot distinguish based on the interval itself.